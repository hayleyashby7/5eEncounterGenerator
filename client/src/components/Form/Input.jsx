/* eslint-disable tailwindcss/no-custom-classname */
/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';

function Input({ label, name, type, register, ...rest }) {
    return (
        <label className='flex flex-1 flex-col' htmlFor={name} aria-labelledby={name}>
            {label}
            <input
                onChange={rest.onChange}
                className='form-input solid rounded  border-2 border-neutral-900 bg-orange-100 px-2 py-1 outline-none'
                id={name}
                type={type}
                {...register(name, { ...rest })}
            />
        </label>
    );
}

export default Input;

Input.propTypes = {
    label: PropTypes.string,
    name: PropTypes.string,
    type: PropTypes.string,
    register: PropTypes.func,
    rest: PropTypes.object,
};
Input.defaultProps = {
    label: '',
    name: '',
    type: '',
    register: () => {},
    rest: {},
};
