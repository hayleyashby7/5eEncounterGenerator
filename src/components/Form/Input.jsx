/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';

function Input({ label, name, type, register, ...rest }) {
    return (
        <label className='flex flex-1 flex-col' htmlFor={name} aria-labelledby={name}>
            {label}
            <input
                onChange={rest.onChange}
                className='border-2 border-solid border-neutral-900 bg-orange-100 outline-none'
                id={name}
                type={type}
                {...register(name, { ...rest })}
            />
        </label>
    );
}

export default Input;

Input.propTypes = {
    label: PropTypes.string,
    name: PropTypes.string,
    type: PropTypes.string,
    register: PropTypes.func,
    rest: PropTypes.object,
};
Input.defaultProps = {
    label: '',
    name: '',
    type: '',
    register: () => {},
    rest: {},
};
