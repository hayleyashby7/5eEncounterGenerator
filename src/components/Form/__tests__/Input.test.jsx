import { describe, it, expect, vi, beforeEach } from 'vitest';
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Input from '../Input';

describe('Input component', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it('renders the input', () => {
        // Arrange
        const { container } = render(<Input />);

        // Assert
        expect(container).toMatchSnapshot();
    });

    it('should render an input with the label and name passed in', () => {
        // Arrange
        const label = 'Test label';
        const name = 'test-name';

        // Act
        render(<Input label={label} name={name} />);

        // Assert
        expect(screen.getByLabelText(label)).toBeInTheDocument();
    });

    it('should render an input with the correct type', () => {
        // Arrange
        const type = 'text';

        // Act
        render(<Input type={type} />);

        // Assert
        expect(screen.getByRole('textbox')).toHaveAttribute('type', type);
    });

    it('should call the onChange function when the input is changed', async () => {
        // Arrange
        const mockOnChange = vi.fn();
        const type = 'text';
        const label = 'Test label';
        const name = 'test-name';
        const user = userEvent.setup();

        render(<Input type={type} label={label} name={name} onChange={mockOnChange} />);

        // Act
        await user.type(screen.getByLabelText(label), 'Test');

        // Assert
        expect(mockOnChange).toBeCalled();
        expect(mockOnChange).toBeCalledTimes(4);
    });
});
