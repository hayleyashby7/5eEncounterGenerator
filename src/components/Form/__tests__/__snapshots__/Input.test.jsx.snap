// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input component > renders the input 1`] = `
<div>
  <label
    aria-labelledby=""
    class="flex flex-1 flex-col"
    for=""
  >
    <input
      class="border-2 border-solid border-neutral-900 bg-orange-100 outline-none"
      id=""
      type=""
    />
  </label>
</div>
`;
